shell/platform/darwin/ios/ios_surface_metal_impeller.mm

```diff
@@ -35,9 +100,10 @@
 
 // |IOSSurface|
 std::unique_ptr<Surface> IOSSurfaceMetalImpeller::CreateGPUSurface(GrDirectContext*) {
-  return std::make_unique<GPUSurfaceMetalImpeller>(this,              //
-                                                   impeller_context_  //
-
+  auto context = std::make_shared<CustomColorAttachmentPixelFormatContext>(
+      impeller_context_, FromMTLPixelFormat(layer_.get().pixelFormat));
```

Comment:
Since this default attachment format is backend-specific, and different backends have different techniques for resolving them (see `ContextVK`, for example, which queries device capabilities when making the decision), prefer to override this behavior in `ContextMTL::GetColorAttachmentPixelFormat` directly instead of creating a wrapper.

This could be as simple as adding a `color_attachment_pixel_format_` field/setter in `ContextMTL`, overriding `GetColorAttachmentPixelFormat` to return it, and then calling the new setter here with a backend cast:

```cpp
ContextMTL::Cast(*impeller_context_).SetColorAttachmentPixelFormat(
    FromMTLPixelFormat(layer_.get().pixelFormat));
```

Reply:
I can make that change 2.  It does however introduce a non-safe cast and a mutable ivar.  I kind of think the proxy value is better, then trying to refactor things so we can know the proper value at creation later.  Let's chat about it quickly.

Reply:
Oh, looking into this I remember why we can't mutate the context.  Because the context is shared across multiple surfaces and surfaces may have different pixel formats.  So I can't mutate it, I have to make a new one.

Reply:
Discussed offline: This case should be safe because we're working in the iOS surface, and we won't be supporting a non-Metal backend on iOS. I don't have a very strong opinion about which solution we land here, but I think the current solution introduces new code navigation friction (especially if we replicate this solution for other backends). A castless solution would be to add an override field on the abstract `Context` instead.

Reply:
> Oh, looking into this I remember why we can't mutate the context. Because the context is shared across multiple surfaces and surfaces may have different pixel formats. So I can't mutate it, I have to make a new one.

Ah, that makes sense. This might eventually happen if we have multiple windows on different monitors, right? The more I think about this, the more I suspect our best bet is to get rid of this this value from the context altogether and supply it when making the render call per-frame. The renderer already configures the attachments for the root pass using the pixel format on the drawable itself every frame, and we aught to do something similar for the offscreen textures instead of arbitrarily storing this value in the context where it doesn't belong. After all, the embedder is what's responsible for providing the surface every frame.